### PR TITLE
Add tests for type casts and coercions

### PR DESCRIPTION
**Pull request by jasongin**
For https://github.com/nodejs/abi-stable-node/issues/191

It turns out all our type casts and coercions were implemented correctly (since I had checked the V8 behavior when I coded them). This change adds a new `tests/addons-napi/test_conversions` project to validate all the permutations. Also I added some comments in the `node_api.cc` code where it was unclear to reviewers that it was doing the right thing.

### CODE REVIEW

**Comment by jasongin** [comment]
https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-commit-linux-abi/88/

File: test/addons-napi/test_conversions/test_conversions.c [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+  napi_value input;
+  status = napi_get_cb_args(env, info, &input, 1);
+  if (status != napi_ok) goto Exit;
+
+  bool value;
+  status = napi_get_value_bool(env, input, &value);
+  if (status != napi_ok) goto Exit;
+
+  napi_value output;
+  status = napi_create_boolean(env, value, &output);
```
Comment (boingoing) [inline test/addons-napi/test_conversions/test_conversions.c]: This should be using ```napi_get_boolean```, I believe.

**Review by boingoing — APPROVED** [review]
Looks good and very exhaustive. But I think you're going to have to rebase/merge.

**Comment by jasongin** [comment]
Rebased and fixed `napi_get_boolean`.

https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-commit-linux-abi/89/

**Pull request merged**